// ultimatum/ultimatum.js

if ('object' === typeof module && 'function' === typeof require) {
	module.exports = Ultimatum;
	var node = require('../nodegame-client');
	//console.log(node);
	//console.log(module);
}



function Ultimatum () {
	
	this.name = 'Monitor for Ultimatum';
	this.description = 'No Description';
	this.version = '0.1';
	
	this.minPlayers = 2;
	this.maxPlayers = 10;
	
	this.automatic_step = true;
	
	this.init = function() {
		var that = this;
	};
	
	var pregame1 = function(){
		//console.log(node);
		console.log('Pregame');
	};
	
	var instructions = function(){
		console.log('Instructions');
	};
		
	var game = function(){
		var that = this;
		var groups = this.pl.getNGroups((this.pl.size()/2));
		
		var i;
		var g = null;
		for(i=0;i<groups.length;i++){
			g = groups[i];
			//console.log(g);
			console.log('GSIZE: ' + g.size());
			if (g.size() > 1) {
				// Bidder
				var bidder = g.getRandom();
				//console.log(bidder);
				g.remove(bidder.id);				
				node.say('DATA', 'BIDDER', bidder.id);
				
				// Respondent
				var respondent = g.getRandom();
				//console.log(respondent);
				node.say('DATA', 'RESPONDENT', respondent.id);

				// Make each other aware
				node.say('DATA', 'OTHER', bidder.id, respondent.id);
				node.say('DATA', 'OTHER', respondent.id, bidder.id);

			}
			else {
				var solo =  g.getRandom();
				console.log(solo);
				node.say('DATA', 'SOLO',respondent.id);
			}	
		}
		console.log('Game1');
	};
	
	var postgame1 = function(){
		console.log('Postgame');
	};
	
	var endgame1 = function(){
		console.log('*********************');
		console.log(node.memory.dump());
		console.log('*********************');
		console.log('*********************');
		console.log(node.memory.dump(true));
		console.log('*********************');
		console.log('Game ended');
	};
	
	// Assigning Functions to Loops
	
	var pregameloop = {
		1: pregame1
	};
	
	var instructionsloop = {
		1: instructions
	};
	
	var gameloop = { // The different, subsequent phases in each round
		1: game
	};
	
	var postgameloop = {
		1: postgame1
	};
	
	var endgameloop = {
		1: endgame1
	};
	
	
	// Creating the game object
		
	// LOOPS
	this.loops = {
			1: {loop:pregameloop},
			2: {loop:instructionsloop},
			3: {rounds:5, loop:gameloop},
			4: {loop:postgameloop},
			5: {loop:endgameloop}
		};	
}
