// peerreview/pr.js
if ('object' === typeof module && 'function' === typeof require) {
	module.exports = PeerReview;
}

function PeerReview () {
	
	this.name = 'Game Example';
	this.description = 'General Description';
	this.version = '0.2';
	
	this.minPlayers = 2;
	this.maxPlayers = 10;
	
	this.automatic_step = true;
	
	this.init = function() {};
	
	var pregame1 = function(){
		console.log('Pregame');
	};
	
	var instructions = function() {		
		console.log('Instructions');
	};
		
	var creation = function(){
		var that = that;
		console.log('--------------------->creation');
	};
	
	var submission = function(){
		node.memory.dump('./pr.csv', true);
		//console.log(node.memory.get(true));
		console.log('submission');
	};
	
	var evaluation = function(){
		console.log('evaluation');
	};
	
	var dissemination = function(){
		console.log('dissemination');
	};
	
	var postgame1 = function() {
		console.log('Postgame');
	};
	
	var endgame1 = function() {
		node.memory.dump('./pr2.csv', true);
		console.log(node.memory.get(true));
		console.log('Game ended');
	};
	
	var waiting = function(){
		console.log('Waiting');
	};
	
	
	
	// Assigning Functions to Loops
	
	var pregameloop = {
		1: pregame1
	};
	
	var instructionsloop = {
		1: instructions
	};
	
	var gameloop = { // The different, subsequent phases in each round
		1: creation,
		2: submission,
		3: evaluation,
		4: dissemination
	};
	
	var postgameloop = {
		1: postgame1
	};
	
	var endgameloop = {
		1: endgame1
	};
	
	
	// Creating the game object
		
	// LOOPS
	this.loops = {
			1: {loop:pregameloop},
			2: {loop:instructionsloop},
			3: {rounds:10, loop:gameloop},
			4: {loop:postgameloop},
			5: {loop:endgameloop}
		};	
}
